# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [26]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
from newsapi import NewsApiClient
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()
load_dotenv()


%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/joseph/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [27]:
# Read your api key environment variable
api_key = os.getenv("news_api")

In [28]:
# Create a newsapi client
newsapi = NewsApiClient(api_key=api_key)

In [59]:
# Fetch the Bitcoin news articles
btc = newsapi.get_everything(
    q="Bitcoin OR BTC",
    language="en",
    page_size=100,
    sort_by="relevancy"
)
# Print total articles
print(f"Total articles about Bitcoin: {btc['totalResults']}")

Total articles about Bitcoin: 7393


In [55]:
# Fetch the Ethereum news articles
eth = newsapi.get_everything(
    q="Ethereum OR ETH",
    language="en",
    page_size=100,
    sort_by="relevancy"
)
# Print total articles
print(f"Total articles about Ethereum: {eth['totalResults']}")

Total articles about Ethereum: 3539


In [46]:
# Create the Bitcoin sentiment scores DataFrame
btc_sentiments = []

for article in btc["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        btc_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
        
    except AttributeError:
        pass
    
btc_df = pd.DataFrame(btc_sentiments)

# Reorder DataFrame columns
cols = ["date", "text", "compound", "positive", "negative", "neutral"]
btc_df = btc_df[cols]

btc_df.head()

date                                               text  compound  \
0  2022-02-11  Netflix\r\n is making a docuseries about one o...   -0.7096   
1  2022-02-17  Even in cyberspace, the Department of Justice ...    0.7351   
2  2022-02-13  The couple would never flee from the country a...   -0.3400   
3  2022-02-14  If Sundays Super Bowl ads are any indication, ...    0.7783   
4  2022-02-12  The news of one of the biggest money launderin...    0.5859   

   positive  negative  neutral  
0     0.000     0.169    0.831  
1     0.147     0.000    0.853  
2     0.057     0.118    0.825  
3     0.163     0.000    0.837  
4     0.124     0.000    0.876

In [52]:
# Create the Ethereum sentiment scores DataFrame
eth_sentiments = []

for article in eth["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        eth_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
        
    except AttributeError:
        pass
    
eth_df = pd.DataFrame(eth_sentiments)

# Reorder DataFrame columns
cols = ["date", "text", "compound", "positive", "negative", "neutral"]
eth_df = eth_df[cols]

eth_df.head()

date                                               text  compound  \
0  2022-03-01  In February, shit hit the fan in the usual way...   -0.3182   
1  2022-02-17  Technical analysis isnt a perfect tool, but it...   -0.2498   
2  2022-02-16  People in Colorado will be able to pay their s...   -0.1027   
3  2022-02-25  Coinbase reported that the share of trading vo...    0.6705   
4  2022-02-26  Illustration by James Bareham / The Verge\r\n\...   -0.4588   

   positive  negative  neutral  
0     0.059     0.093    0.848  
1     0.000     0.059    0.941  
2     0.000     0.036    0.964  
3     0.188     0.000    0.812  
4     0.000     0.083    0.917

In [61]:
# Describe the Bitcoin Sentiment
btc_df.describe()

compound    positive    negative     neutral
count  100.000000  100.000000  100.000000  100.000000
mean     0.020948    0.068880    0.060380    0.870750
std      0.476757    0.067669    0.072455    0.085378
min     -0.895700    0.000000    0.000000    0.627000
25%     -0.323650    0.000000    0.000000    0.822750
50%      0.000000    0.057000    0.041000    0.867500
75%      0.428550    0.105750    0.096000    0.938250
max      0.907900    0.292000    0.269000    1.000000

In [62]:
# Describe the Ethereum Sentiment
eth_df.describe()

compound    positive    negative     neutral
count  100.000000  100.000000  100.000000  100.000000
mean     0.155828    0.082400    0.043690    0.873920
std      0.444918    0.070975    0.064776    0.086187
min     -0.913600    0.000000    0.000000    0.627000
25%     -0.060325    0.000000    0.000000    0.830250
50%      0.165500    0.071500    0.000000    0.875000
75%      0.510600    0.125000    0.068250    0.941750
max      0.918600    0.290000    0.312000    1.000000

### Questions:

Q: Which coin had the highest mean positive score?

A: Ethereum had a highest mean positive score of 0.082400, vs Bitcoin's 0.068880

Q: Which coin had the highest compound score?

A: Ethereum had the highest compound score with a max of 0.918600, vs Bitcoin's 0.907900

Q. Which coin had the highest positive score?

A: Bitcoin had the highest positive score with a max of 0.292000, vs Ethereum's 0.290000

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [10]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [11]:
# Instantiate the lemmatizer
# YOUR CODE HERE!

# Create a list of stopwords
# YOUR CODE HERE!

# Expand the default stopwords list if necessary
# YOUR CODE HERE!

In [12]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    # Remove the punctuation from text

   
    # Create a tokenized list of the words
    
    
    # Lemmatize words into root words

   
    # Convert the words to lowercase
    
    
    # Remove the stop words
    
    
    return tokens

In [13]:
# Create a new tokens column for Bitcoin
# YOUR CODE HERE!

In [14]:
# Create a new tokens column for Ethereum
# YOUR CODE HERE!

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [15]:
from collections import Counter
from nltk import ngrams

In [16]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---